azkaban-common/src/main/java/azkaban/executor/ExecutionController.java

```diff
@@ -450,57 +452,76 @@ public LogData getExecutionJobLog(final ExecutableFlow exFlow, final String jobI
   }
 
   /**
-   * If the resource manager and job history server urls are configured, fetch the application
-   * id from the job log and then construct the job link url.
+   * If the Resource Manager and Job History server urls are configured, find all the
+   * Hadoop/Spark application ids present in the Azkaban job's log and then construct the url to
+   * job logs in the Hadoop/Spark server for each application id found. Application ids are
+   * returned in the order they appear in the Azkaban job log.
    *
    * @param exFlow The executable flow.
    * @param jobId The job id.
    * @param attempt The job execution attempt.
-   * @return the job link url.
+   * @return The map of (application id, job log url)
    */
   @Override
-  public String getJobLinkUrl(final ExecutableFlow exFlow, final String jobId, final int attempt) {
-    if (!this.azkProps.containsKey(ConfigurationKeys.RESOURCE_MANAGER_JOB_URL) || !this.azkProps
-        .containsKey(ConfigurationKeys.HISTORY_SERVER_JOB_URL) || !this.azkProps
-        .containsKey(ConfigurationKeys.SPARK_HISTORY_SERVER_JOB_URL)) {
-      return null;
+  public Map<String, String> getExternalJobLogUrls(final ExecutableFlow exFlow, final String jobId,
+      final int attempt) {
+
+    final Map<String, String> jobLogUrlsByAppId = new LinkedHashMap<>();
```

Comment:
Why use LinkedHashMap over regular HashMap? Likewise, later in the code, LinkedHashSet is used instead of HashSet.

Reply:
In the UI we want to list application ids in the order they were launched.
We use LinkedHashMap and LinkedHashSet because they maintain insertion order. HashMap and HashSet don't.